### PR TITLE
Add Prometheus exporter for PDO

### PR DESCRIPTION
**Pull request by dpifke**

### CODE REVIEW

**Comment by glensc** [comment]
@dpifke please add section to README how this is supposed to be used.


also, perhaps the endpoint should be protected with a token? is that common in prometheus ecosystem that the resources it polls are public?

example: https://github.com/perftools/xhgui/pull/289

**Comment by dpifke** [comment]
> @dpifke please add section to README how this is supposed to be used.

Partially done.  I'll add a link to an example Grafana dashboard for https://performance.wikimedia.org/xhgui once I've made one.

> also, perhaps the endpoint should be protected with a token? is that common in prometheus ecosystem that the resources it polls are public?

Generally, Prometheus endpoints don't require any authentication.  The information being exposed isn't anything which couldn't already be scraped from the index page.

**Comment by glensc** [comment]
@dpifke you need to rebase against merge conflicts from the upstream

File: README.md
Comment (glensc) [inline README.md:338]: perhaps use `/export/prometheus` path to be similar to some (not yet merged) exporter:
- https://github.com/perftools/xhgui/pull/302/files

Comment (Krinkle) [reply README.md:243]: In fairness, those exporters tend to have a dedicated HTTP host or port just for their Prometheus exporter. I haven't checked them all, but I don't see much if any precedent for how to expose Prometheus from something that is itself a CGI web application.

I suppose there's a number of ways forward here:

1. Integrate the endpoint within the regular XHGui REST paths as `/export/prometheus` and use declare the custom path in Prometheus (is this possible?)
2. Integrate as `/export/prometheus` and document for Nginx and Apache how to configure a second virtualhost or port that exposes `/metrics` proxy-path'ed to this endpoint. This would be most complient with [Prometheus best practices](https://prometheus.io/docs/instrumenting/writing_exporters/) in that it would allow the XHGui exporter to have a unique port in the 9xxx range and avoids clashes with any other path conventions it might introduce later. For example, it recommends that visiting http://myexporter:9xxx/ shows a page that links to `/metrics`.
3. Integrate as `/metrics` and treat the public XHGui app itself as a Prometheus exporter.

Which one of these to follow I'll leave up to @glensc.

**Comment by Krinkle** [comment]
These queries aren't ideal to expose publicly. There is precedent for this with the `/import` and `/delete` endpoints which are much more powerful in comparison and also don't require any authentication.

We currently leave it to site-admins to stub these out at the web server level. I suppose we could do the same here, so perhaps nothing is required to change here but just wanted to point it out in case @glensc has thoughts on that. See also https://github.com/perftools/xhgui/issues/308.

File: README.md
Comment (glensc) [reply README.md:243]: I have no strong feelings about the URL path, so be it `/metrics` for now.

**Review by glensc — APPROVED** [review]

**Comment by glensc** [comment]
merged as 4b25fc6, could not figure out how to push to developer branch (allow changes from maintainer was not enabled?)

**Comment by glensc** [comment]
Is there a limitation of why this can't be implemented with MongoDB? or you were just not interested in implementing it yourself?

ps: I've created an issue to track differences, would be nice to have your input, as you use PDO: https://github.com/perftools/xhgui/issues/320

**Comment by dpifke** [comment]
Nothing PDO-specific, I just didn't have a recent MongoDB instance set up with which to try to implement it.  (MongoDB is infeasible to run at Wikimedia due to non-OSI-approved license and it having been removed from Debian.)  Most of the functionality that's missing w/ the PDO backend is stuff we weren't really using; I'll add some notes to the linked task as to what we've identified.  Our goal is to eventually contribute patches to bring it up to par.

**Pull request closed without merging**
